[0029] Proposal: Add Root Signature related driver options (#320)
llvm/wg-hlsl: proposals/0002-root-signature-in-clang.md

```

Behaviour:

 - If the entry function does not have a root signature attribute, it will use
the one defined by the macro expansion

 - If the entry function has a root signature attribute, it will overwrite to
use the one defined by the macro expansion

Design Notes: Commonly used. Planned in Clang.

#### Target Root Signature Version

Compiles the "entry" root signature, specified by the given macro expansion, to
a DXIL Container with just the (version specific) RTS0 part.

Usage:

Given a defined macro in the source file (`#define RS "CBV(b0)"`)

```
  -T <root signature version> -E <entry root signature>`
  -T rootsig_1_0 -E RS /Fo RS.bin
  -T rootsig_1_1 -E RS /Fo RS.bin
```

Behaviour:

 - Parse and perform syntactic validations of "entry" root signature
 - Perform the 'basic' sub-set of validations that are available without access
to resource bindings
 - Produces a DXIL container with only the serialized "entry" root signature
as the only part

 - Note: It is not possible to use `-rootsig-define` to overwrite which root
signature will be used as the "entry" root signature.

Design Notes: Commonly used. Planned in Clang.

#### Option `-Qstrip-rootsignature`

> Strips the root signature attribute from the entry function.

Reviewer: I find this wording a little confusing. I think something like "Omits the root signature part from the compiled shader" would be clearer.

Usage:

```
  -Qstrip-rootsignature
```

Behaviour:

 - Parse and perform syntactic validations of "entry" root signature
 - Perform the 'basic' sub-set of validations that are available without access
to resource bindings
 - Produces the DXIL container with the root signature part (RTS0) omitted

Design Notes: Usable within .hlsl source files. Planned in Clang unless there is
a signal of it being unused.

Design Notes: Plan to match DXC behaviour directly such that we still do

> the parsing and basic validations of the root signataure.

Reviewer: Should this be mentioned in the description of what the attribute does? Whether or not the root signature is validated and stripped late or ignored completely by the frontend is a fairly important detail here.

Author: Does attribute refer to the root signature function attribute here?

In DXC, the root signature is parsed and does the basic validations for the root signature and is stripped late. The question is if we want to match that behaviour in clang.

#### Option `-setrootsignature`

Overwrite the root signature part (RTS0) with the serialized root signature
given in the file.

Usage:

```
  -setrootsignature <DXIL Container>
```

Behaviour:

 - Insert/overwrite the RTS0 part of the compiled DXIL Container with the RST0
contents of the input container

>  - Does not perform any validation with the input root signature

Reviewer: Does this still validate root signatures that were specified by the attribute (or some other way) and then drop them and overwrite them with the given signature, or does it ignore them? Why doesn't this validate the input root signature?

 - Requires input to be a valid DXIL container with or without an RST0 part

Design Notes: Planned in Clang. If compiling a stand-alone DXIL Container with

> just the RTS0 is useful then this is presumably also required in that workflow.

Reviewer: I think we need to know more about when/why this is used to decide whether to plan this. I'm uncomfortable with the fact that we don't validate this at all.

#### Option `-extractrootsignature`

Specifies to compile the shader as normal but will only output the RTS0 part
into the produced DXIL Container.

Usage:

```
  -extractrootsignature /Fo <DXIL Container>
```

Behaviour:

 - Behaviour is the same as the
[root signature target option](#target-root-signature-version), with the
exception that it will run all validations with the resource bindings in the
shader code

> Design Notes: Not planned in Clang, as it seems shadowed.

Reviewer: What does "it seems shadowed" mean in this context?

Usage:

```
  -force-rootsig-ver rootsig_1_0
  -force-rootsig-ver rootsig_1_1
```

Behaviour:

 - Validation logic that is specific to the version will be followed, as
described above
 - Serialized root signature in DXIL container will be compliant with specified
version

Design Notes: Required for backwards compatibility. Implemented in Clang.

#### Option `-rootsig-define`:

Sets the root signature attribute for the entry function to be the root
signature defined by the given macro expansion.

Usage:

Given a defined macro either provided in the source file
(`#define RS "CBV(b0)"`), or, as a command line define
(`-D RS="CBV(b0)"`)

```
  -rootsig-define RS
```

Behaviour:

 - If the entry function does not have a root signature attribute, it will use
the one defined by the macro expansion

 - If the entry function has a root signature attribute, it will overwrite to
use the one defined by the macro expansion

Design Notes: Commonly used. Planned in Clang.

#### Target Root Signature Version

Compiles the "entry" root signature, specified by the given macro expansion, to
a DXIL Container with just the (version specific) RTS0 part.

Usage:

> Given a defined macro in the source file (`#define RS "CBV(b0)"`)

Reviewer 2: Perhaps we should note that it's not necessary to specify the root signature macro in the source file, since it may be specified with a `-D` option on the command line instead.  DXC supports this, but doesn't accept an empty source file. The source file will not even be used in this case, so it could in theory be omitted, but DXC requires it, and it seems to require that it has at least 4 bytes of content (that's likely due to the BOM detection).

```
  -T <root signature version> -E <entry root signature>`
  -T rootsig_1_0 -E RS /Fo RS.bin
  -T rootsig_1_1 -E RS /Fo RS.bin
```

Behaviour:

 - Parse and perform syntactic validations of "entry" root signature
 - Perform the 'basic' sub-set of validations that are available without access
to resource bindings
 - Produces a DXIL container with only the serialized "entry" root signature
as the only part

 - Note: It is not possible to use `-rootsig-define` to overwrite which root

> signature will be used as the "entry" root signature.

Reviewer 2: Perhaps it should be noted why (uses `-E` instead)

```suggestion
signature will be used as the "entry" root signature, because this is specified
using the `-E` option when compiling directly to a root signature target.
```

Design Notes: Commonly used. Planned in Clang.

#### Option `-Qstrip-rootsignature`

> Omits the root signature part (RTS0) from the produced DXIL Container.

Reviewer 2: It also can be used to remove the root signature from an input DXIL container file, when that container file is specified along with the `-dumpbin` option.

  -T rootsig_1_0 -E RS /Fo RS.bin
  -T rootsig_1_1 -E RS /Fo RS.bin
```

Behaviour:

 - Parse and perform syntactic validations of "entry" root signature
 - Perform the 'basic' sub-set of validations that are available without access
to resource bindings
 - Produces a DXIL container with only the serialized "entry" root signature
as the only part

 - Note: It is not possible to use `-rootsig-define` to overwrite which root
signature will be used as the "entry" root signature.

Design Notes: Commonly used. Planned in Clang.

#### Option `-Qstrip-rootsignature`

Omits the root signature part (RTS0) from the produced DXIL Container.

Usage:

```
  -Qstrip-rootsignature
```

Behaviour:

 - Parse and perform syntactic validations of "entry" root signature
 - Perform the 'basic' sub-set of validations that are available without access
to resource bindings
 - Produces the DXIL container with the RTS0 omitted

Design Notes: Usable within .hlsl source files. Planned in Clang unless there is
a signal of it being unused.

Design Notes: Plan to match DXC behaviour directly such that we still do
the parsing and basic validations of the root signataure.

#### Option `-setrootsignature`

> Overwrite the root signature part (RTS0) with the RTS0 of provided DXIL
> Container.

Reviewer 2: This can be used to override the root signature part used either in the resulting DXIL container when compiling a shader, or by adding or replacing the root signature part of an already compiled shader in a DXIL Container.

```suggestion
Override the root signature part (RTS0) in the DXIL Container produced either
when compiling a shader, or in a container loaded using the `-dumpbin` option.
```

Usage:

```
  -setrootsignature <DXIL Container>
```

Behaviour:

 - Insert/overwrite the RTS0 part of the compiled DXIL Container with the RST0
contents of the input container

>  - Does not perform any validation with RTS0 of the provided DXIL Container

Reviewer 2: This does perform validation of the RTS0 with the shader you're attaching it to (making sure it supplies the required bindings).  Why is this statement here?

>  - Requires input to be a valid DXIL container with or without an RST0 part

Reviewer 2: Which input?  The input container for the `-setrootsignature` option is required to be a valid DXIL container *with* an `RTS0` part.

The shader input (whether compiling from HLSL or loading a DXIL Container) need not have a root signature attached.

If one is present in the shader input, that root signature must still be well-formed, but it will not be checked against the shader for binding compatibility, and it will not be used for the final DXIL Container output.

 - Note: Even when provided, this will parse/validate a root signature

> specified as an attribute (or another above) before dropping it.

Reviewer 2: It will parse and check basic validity, but not check against the shader bindings, right?  Perhaps we should clarify that.

Plus, we could change this design to completely ignore the attribute when overriding the root signature instead (as a divergence from what DXC does).  In any case, we should be consistent between the `-rootsig-define` behavior and the `setrootsignature` behavior here.

Design Notes: Planned in Clang. If compiling a stand-alone DXIL Container with
just the RTS0 is useful then this is presumably also required in that workflow.

#### Option `-extractrootsignature`

Specifies to compile the shader as normal but will only output the RTS0 part
into the produced DXIL Container.

Usage:

```

>   -extractrootsignature /Fo <DXIL Container>

Reviewer 2: I wish there was a better way to structure things to indicate these cases that work both when compiling a shader and when loading a compiled shader in a DXIL Container using `-dumpbin`.

```

Behaviour:

 - Behaviour is the same as the
[root signature target option](#target-root-signature-version), with the
exception that it will run all validations with the resource bindings in the
shader code

Design Notes: Not planned in Clang. This is on the assumption that running the
additional validations doesn't provide additional utility over the root
signature target option and compiling the shader regularly.

#### Option `-verifyrootsignature`

Given a shader defined from a DXIL Container, perform validations with the
root signature part (RTS0) of the other DXIL Container

Usage:

```

>   -dumpbin <Shader DXIL Container> -verifyrootsignatures <Root Signature DXIL Container>

Reviewer 2: ```suggestion
  -dumpbin <Shader DXIL Container> -verifyrootsignature <Root Signature DXIL Container>
```

```

> Note: `dumpbin` is required to load a binary instead of performing compilation

Reviewer 2: This is great, but I wonder if we can't just have a `-dumpbin` section we refer to from each of the other options that can apply to a loaded DXIL container using this option.

Behaviour:

 - Performs validation that require resource binding information from the
input DXIL Container with the provided RTS0.

Design Notes: Not planned in Clang. We don't want to provide non-compilation
options.

Design Notes: Maybe these validations will be part of `setrootsignature`.

#### Option `/Frs`

Specifies to compile the shader as normal but will also output the RTS0 part
into a separate DXIL Container.

Usage:

```
  /Frs <DXIL Container>
```

Behaviour:

 - Behaviour is the same as `-extractrootsignature`. Except that it will still
output the complete DXIL Container as normal, and, produce an additional DXIL
container with just the RTS0 part.

Design Notes: Not planned in Clang. For similar motivations as

> `extractrootsignature`.

Reviewer 2: This option is often used along with the `-Qstrip-rootsignature` option to produce a shader container without the root signature and a separate serialized root signature file to be used separately with the runtime.

So, we aren't planning on providing any ability to extract a root signature into a separate container when compiling a shader?  Nor are we planning on providing the ability to extract a root signature from a shader container that has the RTS0 part?

I think these are both useful for developers.  The alternative is to compile a root signature directly to a root signature target.  But that requires knowing which root signature define goes with a particular entry point for the equivalent functionality (and having used a define in the first place when specifying the attribute, which isn't required).
